### PR TITLE
add negative sampling loss and bpr_max loss

### PR DESCRIPTION
**Pull request by denkuzin**
# Description

## Type of change

- [x] Feature
- [ ] Fix
- [ ] Release

## Checklist

- [x ] I have followed the [CONTRIBUTING](https://github.com/criteo/deepr/blob/master/docs/CONTRIBUTING.rst) guidelines.
- [  ] I have updated the [CHANGELOG](https://github.com/criteo/deepr/blob/master/docs/CHANGELOG.rst).

### CODE REVIEW

File: deepr/layers/losses.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            Negative Sampling  loss
+        """
+        positives, negatives, mask, weights = tensors
+        true_losses = tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.ones_like(positives), logits=positives)
+        sampled_losses = tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.zeros_like(negatives), logits=negatives)
+        # filter the values that correspond to mask
+        negative_number = tf.reduce_sum(mask, -1)
+        sampled_losses = tf.multiply(sampled_losses, mask)
+        sampled_losses = tf.reduce_sum(sampled_losses, axis=2)
+        sampled_losses = sampled_losses / (negative_number + 1e-8)
```
Comment (guillaumegenthial) [inline deepr/layers/losses.py:160]: it might makes sense to use tf.div_no_nan ?

File: deepr/layers/losses.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        """
+        positives, negatives, mask, weights = tensors
+        positives, negatives = make_same_shape([positives, negatives], broadcast=False)
+        no_sampled_logits = tf.cast(tf.greater_equal(tf.reduce_sum(mask, -1), 0), tf.float32)
+        softmax_scores = self._compute_softmax_scores(negatives, mask)
+        # compute bpr_max losses
+        losses = -tf.multiply(
+            no_sampled_logits,
+            tf.log(
+                tf.reduce_sum(tf.multiply(tf.multiply(softmax_scores, tf.nn.sigmoid(positives - negatives)), mask), 2)
```
[the remark this answers is not in the document]
Comment (denkuzin) [reply deepr/layers/losses.py:85]: yes, good point

File: deepr/layers/losses.py (hunk 1 above)
Comment (denkuzin) [reply deepr/layers/losses.py:160]: yes, it's better, agree

**Comment by denkuzin** [comment]
yes, the tests will be in a separate review
